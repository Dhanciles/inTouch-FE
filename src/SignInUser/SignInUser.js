import React, { Component } from 'react'; 
import './SignInUser.css'
import { authenticateUser } from '../Thunks/authenticateUser.js'
import { isLoading, hasErrored, signInUser } from '../actions/index.js'
import { connect } from 'react-redux'

export default class SignInUser extends Component {
  constructor() {
    super()
    this.state = {
      userName: '', 
      password: '',
      disabled: true
    }
  }
  
  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({
      [name]: value
    }, () => this.enableButton())
  }
  
  enableButton = () => {
    const { userName, password } = this.state
    if (userName !== '' && password  !== '') {
      this.setState({
        disabled: false
      })
    }
  }
  
  render() {
    const { userName, password, disabled } = this.state
    return (
      <form onSubmit={ this.handleSubmit }>
          <input onChange={ this.handleChange } className="user-name" name="userName" value={userName} type="text" required placeholder="Username"/>
          <input onChange={ this.handleChange } className="user-password" name="password" value={password} type="password" required placeholder="Password"/>
        <button className="sign-in-button" disabled={disabled}>Sign In</button>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  signIn: (userName, password) => dispatch(signInUser(userName, password))
})

export default connect(null, mapDispatchToProps)(SignInUser)