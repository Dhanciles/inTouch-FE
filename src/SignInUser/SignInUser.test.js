import React from 'react'; 
import { shallow } from 'enzyme'; 
import SignInUser from './SignInUser.js'; 

describe('SignInUser', () => {
  let mockName; 
  let mockPassword;
  let mockEvent
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<SignInUser signIn={jest.fn()}/>)
    mockName = 'Rajaa'; 
    mockPassword = 'password'; 
    mockEvent = {
     target: {
       name: 'token',
       value: 'token12345'
      }
    }
  })

  it('should match snapshot', () => {
    const wrapper = shallow(<SignInUser />)

    expect(wrapper).toMatchSnapshot()
  });
  
  describe('handleChange', () => {
    let mockEvent
    let wrapper
    
    beforeEach(() => {
      wrapper = shallow(<SignInUser signIn={jest.fn()}/>)
      mockEvent = {
       target: {
         name: 'token',
         value: 'token12345'
        }
      }
    })
    
    it('should invoke handleChange when event occurs on an input', () => {
      wrapper.handleChange = jest.fn()
      wrapper.find('.user-name').simulate('change', mockEvent)
      
      expect(wrapper.handleChange).toHaveBeenCalled
    })
      
    it('should update state when handleChange is invoked on an input', () => {
      const expected = {
        userName: 'Derek',  
        password: '',
        disabled: true
      }
      
      wrapper.instance().handleChange(mockEvent)
      
      expect(wrapper.state()).toEqual(expected)
    })
    
    it('should invoke enableButton', () => {
      const mockFunction = jest.fn()
      wrapper.enableButton = mockFunction
      
      wrapper.instance().handleChange(mockEvent)
      
      expect(mockFunction).toHaveBeenCalled
    })
  })
  
  describe('enableButton', () => {
    let wrapper
    
    beforeEach(() => {
      wrapper = shallow(<SignInUser signIn={jest.fn()}/>)
    })

    it('should toggle disabled if all values are present in input fields', () => {
      const mockState = {
        userName: 'Derek', 
        password: 'password',
        disabled: false
      }    
      const expected = {
        userName: 'Derek', 
        password: 'password',
        disabled: false
      }
      
      wrapper.setState(mockState)
      wrapper.instance().enableButton()
      
      expect(wrapper.state()).toEqual(expected)
    })
  })
  
  describe('handleSubmit', () => {
    let wrapper
    let mockSubmitEvent
    
    beforeEach(() => {
      wrapper = shallow(<SignInUser signIn={jest.fn()}/>)
    })
    
    it('should invoke handleSubmit on click of the Sign In button', () => {
      mockSubmitEvent = {
        target: {},
        preventDefault: () => {}
      }
      
      wrapper.setState({
        disabled: false
      })
      
      wrapper.find('.sign-in-button').simulate('click', mockSubmitEvent)
      
      expect(wrapper.handleSubmit).toHaveBeenCalled
    })
  })
});
