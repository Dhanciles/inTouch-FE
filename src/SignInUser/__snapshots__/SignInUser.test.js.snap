// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInUser should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignInUser
    signIn={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="user-name"
          name="userName"
          onChange={[Function]}
          placeholder="Username"
          required={true}
          type="text"
          value=""
        />,
        <input
          className="user-password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />,
        <button
          className="sign-in-button"
          disabled={true}
        >
          Sign In
        </button>,
        <p>
          New User?
        </p>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "user-name",
          "name": "userName",
          "onChange": [Function],
          "placeholder": "Username",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "user-password",
          "name": "password",
          "onChange": [Function],
          "placeholder": "Password",
          "required": true,
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign In",
          "className": "sign-in-button",
          "disabled": true,
        },
        "ref": null,
        "rendered": "Sign In",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "New User?",
        },
        "ref": null,
        "rendered": "New User?",
        "type": "p",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="user-name"
            name="userName"
            onChange={[Function]}
            placeholder="Username"
            required={true}
            type="text"
            value=""
          />,
          <input
            className="user-password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
          />,
          <button
            className="sign-in-button"
            disabled={true}
          >
            Sign In
          </button>,
          <p>
            New User?
          </p>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "user-name",
            "name": "userName",
            "onChange": [Function],
            "placeholder": "Username",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "user-password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "Password",
            "required": true,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign In",
            "className": "sign-in-button",
            "disabled": true,
          },
          "ref": null,
          "rendered": "Sign In",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "New User?",
          },
          "ref": null,
          "rendered": "New User?",
          "type": "p",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
